src/firefly/js/drawingLayers/WebGrid.js

```diff
@@ -0,0 +1,882 @@
+ /*
+ * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
+ * Lijun
+ * 4/14/16
+ */
+
+import  {visRoot} from '../visualize/ImagePlotCntlr.js';
+import {makeDrawingDef} from '../visualize/draw/DrawingDef.js';
+import DrawLayer, {ColorChangeType}  from '../visualize/draw/DrawLayer.js';
+import CsysConverter from '../visualize/CsysConverter.js';
+import { makeWorldPt, makeImagePt,makeImageWorkSpacePt} from '../visualize/Point.js';
+import ShapeDataObj from '../visualize/draw/ShapeDataObj.js';
+import {primePlot} from '../visualize/PlotViewUtil.js';
+import {makeFactoryDef} from '../visualize/draw/DrawLayerFactory.js';
+import CoordinateSys from '../visualize/CoordSys.js';
+import CoordUtil from '../visualize/CoordUtil.js';
+import {getUIComponent} from './WebGridUI.jsx';
+import {get} from 'lodash';
+ import AppDataCntlr from '../core/AppDataCntlr.js';
+
+export const COORDIANTE_PREFERENCE = 'coordinate';
+
+const  coordinateArray = [
+     {coordName:'eq2000hms',        csys:CoordinateSys.EQ_J2000},
+     {coordName:'eq2000dcm',        csys:CoordinateSys.EQ_J2000},
+     {coordName:'eqb1950hms',       csys:CoordinateSys.EQ_B1950},
+     {coordName:'eqb1950dcm',       csys: CoordinateSys.EQ_B1950},
+     {coordName:'galactic',         csys:CoordinateSys.GALACTIC},
+     {coordName:'superGalactic',    csys:CoordinateSys.SUPERGALACTIC},
+     {coordName:'epj2000',          csys:CoordinateSys.ECL_J2000},
+     {coordName:'epb1950',          csys:CoordinateSys.ECL_B1950}
+
+ ];
+
+const ID= 'WEB_GRID';
+const TYPE_ID= 'WEB_GRID_TYPE';
+const factoryDef= makeFactoryDef(TYPE_ID,creator, getDrawData,getLayerChanges, null,getUIComponent);
+export default {factoryDef, TYPE_ID}; // every draw layer must default export with factoryDef and TYPE_ID
+const RANGE_THRESHOLD = 1.02;
+/**
+ * Color used to draw the grid
+ **/
+const  DEF_GRID_COLOR = 'green';
+const minUserDistance= 0.25;   // user defined max dist. (deg)
+const maxUserDistance= 3.00;   // user defined min dist. (deg)
+
+var useLabels= true;
+var userDefinedDistance = false;
+var idCnt=0;
+
+
+/**
+ *
+ * @return {Function}
+ */
+function creator() {
+
+    var drawingDef= makeDrawingDef( DEF_GRID_COLOR);
+    var options= {
+        hasPerPlotData:true,
+        isPointData:false,
+        canUserChangeColor: ColorChangeType.DYNAMIC
+    };
+
+    return DrawLayer.makeDrawLayer( `${ID}-${idCnt}`, TYPE_ID, 'web grid',
+        options , drawingDef, {});
+
+}
+
+ /**
+  * Implement this method defined in DrawLayerFactory.js
+  * @param dataType
+  * @param plotId
+  * @param drawLayer
+  * @param action
+  * @param lastDataRet
+  * @returns {*}
+  */
+
+function getDrawData(dataType, plotId, drawLayer, action, lastDataRet){
+
+     var plot= primePlot(visRoot(),plotId);
+     if (!plot)return null;
+
+     var cc= CsysConverter.make(plot);
+     if (!cc) return null;
+
+     var drawDataArray= getData(plot, cc);
+     return drawDataArray|| lastDataRet;
```

Comment:
Would be better to do `lastDataRet || getData(plot,cc)` that way you don't redraw unnecessarily.  If you return `lastDataRet` the object does not update and then it will not redraw.

To make this all work right you will have use `getLayerChanges`. When there is a coordinate system change (because you are now using the `CUSTOM_DATA` action) you null out the drawing data.

_Like this:_
```
function getLayerChanges(drawLayer, action) {
    if  (action.type!==DrawLayerCntlr.MODIFY_CUSTOM_FIELD) return null; // don't do anything
    const {coordSys}= action.payload;                  // get new coordSys
    if (coordSys!==drawLayer.coorSys) return null;       // it not different, don't do anything
    const dd= Object.assign({},drawLayer.drawData);   // the the drawData from the layer
    dd[DataTypes.DATA]= null;                                      // clear it
    return Object.assign{coordSys},{drawData:dd});     // return the drawLayer changes
}
```

Now the last line of `getDrawData` can be:
```
return lastDataRet || getData(plot,cc);
```